File: app/src/test/java/no/nav/tilbakemeldingsmottak/itest/BestillingAvSamtaleIT.java
package no.nav.tilbakemeldingsmottak.itest;

import no.nav.tilbakemeldingsmottak.rest.bestillingavsamtale.domain.BestillSamtaleRequest;
import no.nav.tilbakemeldingsmottak.rest.bestillingavsamtale.domain.BestillSamtaleResponse;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

import static no.nav.tilbakemeldingsmottak.TestUtils.createBestillSamtaleRequest;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class BestillingAvSamtaleIT extends ApplicationTest {

    private static final String URL_BESTILLING_AV_SAMTALE = "/rest/bestilling-av-samtale";

    @Test
    void happyPath()  {
        BestillSamtaleRequest request = createBestillSamtaleRequest();
        HttpEntity<BestillSamtaleRequest> requestEntity = new HttpEntity<>(request, createHeaders());
        ResponseEntity<BestillSamtaleResponse> response = restTemplate.exchange(URL_BESTILLING_AV_SAMTALE, HttpMethod.POST, requestEntity, BestillSamtaleResponse.class);

        assertEquals(HttpStatus.OK, response.getStatusCode());
    }
}

File: app/src/test/kotlin/no/nav/tilbakemeldingsmottak/itest/BestillingAvSamtaleIT.kt
package no.nav.tilbakemeldingsmottak.itest

import no.nav.tilbakemeldingsmottak.ApplicationTest
import no.nav.tilbakemeldingsmottak.config.Constants
import no.nav.tilbakemeldingsmottak.util.builders.BestillSamtaleRequestBuilder
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.`when`
import org.springframework.beans.factory.annotation.Value

internal class BestillingAvSamtaleIT : ApplicationTest() {

    private val URL_BESTILLING_AV_SAMTALE = "/rest/bestilling-av-samtale"

    @Value("\${auth.issuers.tokenx.issuer-uri}")
    lateinit var tokenxIssuer: String

    val tilbakemeldinger = "tilbakemeldinger"

    @Test
    fun `happy path`() {
        // Given
        val request = BestillSamtaleRequestBuilder().build()
        val mockJwt = createMockJwt(tokenxIssuer)

        `when`(tokenxJwtDecoder.decode(anyString())).thenReturn(mockJwt)

        // When / Then
        restTemplate!!.post()
            .uri(URL_BESTILLING_AV_SAMTALE)
            .headers { it.addAll(createHeaders(Constants.TOKENX_ISSUER, tilbakemeldinger)) }
            .bodyValue(request)
            .exchange()
            .expectStatus().is2xxSuccessful
    }


    @Test
    fun `validation error, illegal telephone number`() {
        // Given
        val request = BestillSamtaleRequestBuilder().build(telefonnummer = "ABC-9999")
        val requestEntity = HttpEntity<BestillSamtaleRequest>(request, createHeaders())

        // When
        val response: ResponseEntity<BestillSamtaleResponse> = restTemplate!!.exchange(
            URL_BESTILLING_AV_SAMTALE, HttpMethod.POST, requestEntity, BestillSamtaleResponse::class.java
        )

        // Then
        assertEquals(HttpStatus.BAD_REQUEST, response.statusCode)
    }


    @Test
    fun `validation error, periode ikke angitt`() {
        // Given
        val request = BestillSamtaleRequestBuilder().build(tidsrom = null)
        val requestEntity = HttpEntity<BestillSamtaleRequest>(request, createHeaders())

        // When
        val response: ResponseEntity<BestillSamtaleResponse> = restTemplate!!.exchange(
            URL_BESTILLING_AV_SAMTALE, HttpMethod.POST, requestEntity, BestillSamtaleResponse::class.java
        )

        // Then
        assertEquals(HttpStatus.BAD_REQUEST, response.statusCode)
    }


}
